src/SIMD.jl

```diff
@@ -1593,4 +1614,115 @@ end
     end
 end
 
+export VecRange
+
+"""
+    VecRange{N}(i::Int)
+
+Analogous to `UnitRange` but for loading SIMD vector of width `N` at
+index `i`.
+
+# Examples
+```jldoctest
+julia> xs = ones(4);
+
+julia> xs[VecRange{4}(1)]  # calls `vload(Vec{4,Float64}, xs, 1)`
+<4 x Float64>[1.0, 1.0, 1.0, 1.0]
+```
+"""
+struct VecRange{N}
+    i::Int
+end
+
+@inline Base.length(idx::VecRange{N}) where {N} = N
+@inline Base.first(idx::VecRange) = idx.i
+@inline Base.last(idx::VecRange) = idx.i + length(idx) - 1
+
+@inline Base.:+(idx::VecRange{N}, j::Integer) where N = VecRange{N}(idx.i + j)
+@inline Base.:+(j::Integer, idx::VecRange{N}) where N = VecRange{N}(idx.i + j)
+@inline Base.:-(idx::VecRange{N}, j::Integer) where N = VecRange{N}(idx.i - j)
+
+Base.checkindex(::Type{Bool}, inds::AbstractUnitRange, idx::VecRange) =
+    (first(inds) <= first(idx)) && (last(idx) <= last(inds))
+
+Base.checkindex(::Type{Bool}, inds::AbstractUnitRange, idx::Vec) =
+    all(first(inds) <= idx) && all(idx <= last(inds))
+
+@inline _checkarity(::AbstractArray{<:Any,N}, ::Vararg{<:Any,N}) where {N} =
+    nothing
+
+@inline _checkarity(::AbstractArray, ::Any) = nothing
+
+_checkarity(::AbstractArray{<:Any,N}, ::Vararg{<:Any,M}) where {N,M} =
+    throw(ArgumentError("""
+    $M indices are given to $N-dimensional array.
+    Exactly $N (non-mask) indices have to be specified when using SIMD.
+    """))
+
+# Combined with `_preprocessindices`, helper function `_extractmask`
+# extracts `mask` in the tail position.  As slicing tuple is not
+# type-stable, we use reverse-of-tail-of-reverse hack to extract
+# `mask` at the end of `args`.
+@inline _extractmask(mask::Vec{N,Bool}, R::Vararg{Integer}) where N =
+    (reverse(R), mask)
+@inline _extractmask(R::Vararg{Integer}) = (reverse(R), nothing)
+@inline _extractmask(mask::Vec{N,Bool}) where {N} = ((), mask)
+@inline _extractmask() = ((), nothing)
+
+@noinline _extractmask(rargs...) =
+    throw(ArgumentError("""
+    Using SIMD indexing `array[idx, i2, ..., iN, mask]` for `N`-dimensional
+    array requires `i2` to `iN` to be all integers and `mask` to be optionally
+    a SIMD vector `Vec` of `Bool`s.  Given `(i2, ..., iN, mask)` is
+    $(summary(reverse(rargs)))
+    """))
+
+_maskedidx(idx, ::Nothing, ::Any) = idx
+_maskedidx(idx::Vec, mask::Vec, fst) = vifelse(mask, idx, fst)
+_maskedidx(idx::VecRange, mask::Vec, fst) =
+    _maskedidx(Vec(ntuple(i -> i - 1 + idx.i, length(mask))), mask, fst)
+
+Base.@propagate_inbounds function _preprocessindices(arr, idx, args)
+    I, mask = _extractmask(reverse(args)...)
+    _checkarity(arr, idx, I...)
+    @boundscheck checkbounds(arr,
+                             _maskedidx(idx, mask, first(axes(arr, 1))),
+                             I...)
+    return I, mask
+end
+
+Base.@propagate_inbounds function Base.getindex(
+        arr::Union{Array{T},SubArray{T}}, idx::VecRange{N},
```

Comment:
These `getindex` methods cannot be used generally on all `SubArray`s — the use of `pointer` means that they need to at least be `StridedSubArray`s.  This will fail for a `view(A, [1,2], [1,2])`, for example.

Further, am I correct in understanding that `vload` will load the next `N` contiguous elements after the pointer?  That further restricts us to `Base.FastContiguousSubArray`s.

Reply:
Thanks, that's exactly what @timholy just commented in https://github.com/JuliaArrays/MappedArrays.jl/pull/24#issuecomment-460568978 :)

I think we need to use `Base.FastContiguousSubArray` in various places in SIMD.jl.  Also:

> I think the best way to do it is to use `vlolad`/`vstore` when it's continuous and `vgather`/`vscatter` when it's not.

It also would be nice to support `ReinterpretArray` but that may require to fix some assumption on alignment.

Reply:
Ah, of course he does! Sorry, I was reading these simultaneously and hadn't gotten to his comment there yet. :)

Reply:
I appreciate that you pointed it out.  It's fixed in #44.